services/galley/src/Galley/API/Message.hs

```diff
@@ -0,0 +1,332 @@
+      }
+
+newMessagePush ::
+  Ord k =>
+  Domain ->
+  Map k LocalMember ->
+  Maybe ConnId ->
+  MessageMetadata ->
+  (k, ClientId) ->
+  Event ->
+  MessagePush
+newMessagePush localDomain members mconn mm (k, client) e = fromMaybe mempty $ do
+  member <- Map.lookup k members
+  newBotMessagePush member <|> newUserMessagePush member
+  where
+    newBotMessagePush :: LocalMember -> Maybe MessagePush
+    newBotMessagePush member = newBotPush <$> newBotMember member <*> pure e
+    newUserMessagePush :: LocalMember -> Maybe MessagePush
+    newUserMessagePush member =
+      fmap newUserPush $
+        newConversationEventPush localDomain e [memId member]
+          <&> set pushConn mconn
+            . set pushNativePriority (mmNativePriority mm)
+            . set pushRoute (bool RouteDirect RouteAny (mmNativePush mm))
+            . set pushTransient (mmTransient mm)
+            . set (pushRecipients . traverse . recipientClients) (RecipientClientsSome (singleton client))
+
+data MessageMetadata = MessageMetadata
+  { mmNativePush :: Bool,
+    mmTransient :: Bool,
+    mmNativePriority :: Maybe Priority,
+    mmData :: ByteString
+  }
+  deriving (Eq, Ord, Show)
+
+qualifiedNewOtrMetadata :: Public.QualifiedNewOtrMessage -> MessageMetadata
+qualifiedNewOtrMetadata msg =
+  MessageMetadata
+    { mmNativePush = Public.qualifiedNewOtrNativePush msg,
+      mmTransient = Public.qualifiedNewOtrTransient msg,
+      mmNativePriority = Public.qualifiedNewOtrNativePriority msg,
+      mmData = Public.qualifiedNewOtrData msg
+    }
+
+-- unqualified
+
+legacyClientMismatchStrategy :: Domain -> Maybe [UserId] -> Maybe Public.IgnoreMissing -> Maybe Public.ReportMissing -> ClientMismatchStrategy
+legacyClientMismatchStrategy localDomain (Just uids) _ _ =
+  MismatchReportOnly (Set.fromList (map (`Qualified` localDomain) uids))
+legacyClientMismatchStrategy _ Nothing (Just IgnoreMissingAll) _ = MismatchIgnoreAll
+legacyClientMismatchStrategy localDomain Nothing (Just (IgnoreMissingList uids)) _ =
+  MismatchIgnoreOnly (Set.map (`Qualified` localDomain) uids)
+legacyClientMismatchStrategy _ Nothing Nothing (Just ReportMissingAll) = MismatchReportAll
+legacyClientMismatchStrategy localDomain Nothing Nothing (Just (ReportMissingList uids)) =
+  MismatchReportOnly (Set.map (`Qualified` localDomain) uids)
+legacyClientMismatchStrategy _ Nothing Nothing Nothing = MismatchReportAll
+
+legacyMismatch :: Domain -> MessageSendingStatus -> ClientMismatch
+legacyMismatch localDomain status =
+  ClientMismatch
+    { cmismatchTime = mssTime status,
+      missingClients = unqualify (mssMissingClients status),
+      redundantClients = unqualify (mssRedundantClients status),
+      deletedClients = unqualify (mssDeletedClients status)
+    }
+  where
+    unqualify = UserClients . Map.findWithDefault mempty localDomain . qualifiedUserClients
+
+unqualifiedResponse :: Domain -> Union PostOtrResponses -> Union PostOtrResponsesUnqualified
```

Comment:
Unfortunately, `respond` isn't good enough, because it only gives you something of the form `a -> NS as` where `a` is a member of `as`. What we need here is `NS as -> NS bs`, given a way to map the types inside `as` to the types inside `bs`.

After going through the docs for SOP, I've found the `htrans` function, which seems to be general enough for this. It works using an arbitrary type-class which is able to convert `as` into `bs`. For this case, I've created an `Unqualify a b` type class which knows how to convert `a` into `b` given the local domain. Then after giving instances for all the types involved, including a trivial one for `Unqualify a a`, the conversion between the two unions can then be simply implemented in terms of `htrans`.